File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/repository/HomeSnapshotStore.kt
package com.rork.guidestreamtvandroid.data.repository

import android.content.Context
import com.rork.guidestreamtvandroid.data.models.TMDBResult
import com.rork.guidestreamtvandroid.data.remote.RecommendedTitle
import com.rork.guidestreamtvandroid.data.remote.StreamingReleasesService
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File

/**
 * Last-known Home rails, persisted so the next launch paints real content
 * immediately instead of a shimmer while the cold load completes. Only the
 * TMDB-backed rails, the streaming_releases rows and the resolved provider
 * names are stored — live sports, creator uploads and anything time-sensitive
 * are refetched. Mirrors iOS HomeSnapshotStore.swift.
 */
@Serializable
data class HomeSnapshot(
    val version: Int = HomeSnapshot.CURRENT_VERSION,
    val savedAtMs: Long,
    val trending: List<TMDBResult>,
    val onAir: List<TMDBResult>,
    val topRated: List<TMDBResult>,
    val genreShows: List<TMDBResult>,
    val bingeReady: List<TMDBResult>,
    val releaseRows: List<StreamingReleasesService.StreamingReleaseRow>,
    /** Recommended for You — the first rail under the hero; without it the
     * snapshot paints Today's Pick first and the page reflows when it lands. */
    val recommendedTitles: List<RecommendedTitle> = emptyList(),
    /** tmdbId → provider display name, rebuilt into Platform on load. */
    val providerNames: Map<Int, String>,
) {
    companion object { const val CURRENT_VERSION = 2 }
}

object HomeSnapshotStore {
    /** Snapshots older than this are ignored rather than shown stale. */
    private const val MAX_AGE_MS = 24L * 60 * 60 * 1000

    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    @Volatile private var file: File? = null

    fun init(context: Context) {
        file = File(context.applicationContext.cacheDir, "home_snapshot_v${HomeSnapshot.CURRENT_VERSION}.json")
    }

    fun load(): HomeSnapshot? {
        val f = file ?: return null
        return try {
            if (!f.exists()) return null
            val snap = json.decodeFromString(HomeSnapshot.serializer(), f.readText())
            if (snap.version != HomeSnapshot.CURRENT_VERSION) return null
            if (System.currentTimeMillis() - snap.savedAtMs > MAX_AGE_MS) return null
            if (snap.trending.isEmpty()) return null
            snap
        } catch (_: Exception) {
            null
        }
    }

    fun save(snapshot: HomeSnapshot) {
        val f = file ?: return
        try {
            val tmp = File(f.parentFile, f.name + ".tmp")
            tmp.writeText(json.encodeToString(HomeSnapshot.serializer(), snapshot))
            if (!tmp.renameTo(f)) { f.writeText(tmp.readText()); tmp.delete() }
        } catch (_: Exception) {}
    }

    fun clear() {
        try { file?.delete() } catch (_: Exception) {}
    }
}

File: ios/GuideStreamTV/Services/HomeSnapshotStore.swift
import Foundation

/// Last-known Home rails, persisted so the next launch paints real content
/// immediately instead of a shimmer while ~80 requests complete. Only the
/// TMDB-backed rails and their resolved provider names are stored — live
/// sports, creator uploads and anything time-sensitive are refetched.
nonisolated struct HomeSnapshot: Codable, Sendable {
    static let currentVersion = 2

    var version: Int = HomeSnapshot.currentVersion
    var savedAt: Date
    var trending: [TMDBResult]
    var onAir: [TMDBResult]
    var bingeFallback: [TMDBResult]
    var newToday: [TMDBResult]
    var topRated: [TMDBResult]
    var genreShows: [TMDBResult]
    var recommendedShows: [TMDBResult]
    /// Recommended for You — the first rail under the hero. Without it the
    /// snapshot paints Today's Pick first and the page reflows when it lands.
    var recommendedTitles: [RecommendedTitle]
    var newReleases: [StreamingRelease]
    /// tmdbId → provider display name, rebuilt into `Platform` on load.
    var providerNames: [Int: String]
}

/// Disk store for `HomeSnapshot`. Lives in Caches so the OS may evict it;
/// a missing or unreadable file simply means the normal cold load runs.
nonisolated enum HomeSnapshotStore {
    /// Snapshots older than this are ignored rather than shown stale.
    static let maxAge: TimeInterval = 24 * 60 * 60

    private static var fileURL: URL? {
        FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?
            .appendingPathComponent("home_snapshot_v\(HomeSnapshot.currentVersion).json")
    }

    static func load() -> HomeSnapshot? {
        guard let url = fileURL, let data = try? Data(contentsOf: url) else { return nil }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        guard let snap = try? decoder.decode(HomeSnapshot.self, from: data) else { return nil }
        guard snap.version == HomeSnapshot.currentVersion,
              Date().timeIntervalSince(snap.savedAt) < maxAge,
              !snap.trending.isEmpty else { return nil }
        return snap
    }

    static func save(_ snapshot: HomeSnapshot) {
        guard let url = fileURL else { return }
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(snapshot) else { return }
        try? data.write(to: url, options: .atomic)
    }

    static func clear() {
        guard let url = fileURL else { return }
        try? FileManager.default.removeItem(at: url)
    }
}
